### PR TITLE
Plugable config scope

### PR DESCRIPTION
**Pull request by arj03**
This one should be ready as well. It is an update to #21 and implements the multiserver 1.13 changes.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -142,57 +94,77 @@ module.exports = function (opts) {
...
+              transports.forEach(function (transport) {
+                if (transport.name == outTransportType && transform.name == conf.transform)
+                  client_suites.push([
+                    transport.create(conf),
+                    transform.create()
+                  ])
+              })
+            })
+          })
+        }
```
Comment (christianbundy) [inline index.js]: Logic looks great. Do you know which versions of JavaScript we're supporting? If we can use `Object.values()` we can replace the top two loops (here and elsewhere) with:

```js
const confs = Object.values(opts.connections.outgoing);
Object.values(confs).forEach(function (conf) {
```

Not sure whether it's in scope for this PR, but we could also use a Cartesian product and replace *all* of the nested loops with something like:

```js
var confs = Object.values(opts.connections.outgoing)
cartesian(confs, transforms, transports).forEach(function (conf, transform, transport) {
  if (transport.name == outTransportType && transform.name == conf.transform)
    client_suites.push([
      transport.create(conf),
      transform.create()
    ])
})
```

I think lodash and friends can do Cartesian products, but I [found](https://stackoverflow.com/a/43053803) a two-liner (in ES6):

```js
const f = (a, b) => [].concat(...a.map(d => b.map(e => [].concat(d, e))));
const cartesian = (a, b, ...c) => (b ? cartesian(f(a, b), ...c) : a);
```

Comment (cryptix) [reply index.js]: I like the less nested and shorter code but that two-liner also isn't easy to understand but a clear name and a comment should help with that, I think...

AFAICT `Object.values` works in node 8. I'm not sure about 6 and how long we want to support it.

Comment (dominictarr) [reply index.js]: @christianbundy backend stuff tries to target es5

Comment (cryptix) [reply index.js]: from the SO link:

> My example uses the spread operator and the rest parameters

According to the mozilla MDN, spread is supported since node 5 and rest param since node 6. So I think this would be fine to since we only test node v6 and upwards anyway?

@christianbundy can you link me to the _elsewhere_ you mentioned? Depending on the number of cases we might want to do this in a separate PR?

File: index.js [hunk 2]
```diff
@@ -142,57 +94,77 @@ module.exports = function (opts) {
...
+          opts.connections.incoming[incTransportType].forEach(function (conf) {
+            transforms.forEach(function (transform) {
+              transports.forEach(function (transport) {
+                if (transport.name == incTransportType && transform.name == conf.transform)
+                  server_suites.push([
+                    transport.create(conf),
+                    transform.create()
+                  ])
+              })
+            })
```
Comment (christianbundy) [inline index.js]: @cryptix Here's the other one I mentioned. We can either do it here or in a separate PR, either way works great for me. Also instead of specifying `cartesian` and such I'm sure there's a library we can use. Maybe [lodash.product](https://www.npmjs.com/package/lodash.product)?

File: index.js [hunk 3]
```diff
@@ -88,52 +69,23 @@ module.exports = function (opts) {
...
+  return create
+  .use({
     manifest: {
       auth: 'async',
       address: 'sync',
       manifest: 'sync',
     },
     init: function (api, opts, permissions, manifest) {
 
+      //legacy, but needed to pass tests.
```
[the remark this answers is not in the document]
Comment (arj03) [reply index.js]: I think instead of calling it *// needed for tests*, we should just call it sane defaults ;-)

File: index.js [hunk 4]
```diff
@@ -227,11 +200,22 @@ module.exports = function (opts) {
...
           })
         },
 
+        multiserver: {
+          transport: function (transport) {
+            if(server) throw new Error('cannot add protocol after server initialized')
+            transports.push(transport); return this
+          },
+          transform: function (transform) { transforms.push(transform); return this },
+          parse: function (str) { // FIXME: server / client? or try both?
```
[the remark this answers is not in the document]
Comment (arj03) [reply index.js]: I did. I added it so that I would come back to it and fix it before opening a PR :-) Thanks for spotting this!

**Comment by arj03** [comment]
Thanks for the review @cryptix. I pushed two more commits to address the things you mentioned.

File: index.js [hunk 5]
```diff
@@ -78,7 +58,8 @@ function msLogger (stream) {
 
 //opts must have appKey
 module.exports = function (opts) {
-
+  //this weird thing were some config is loaded first, then the rest later... not necessary.
```
[the remark this answers is not in the document]
Comment (dominictarr) [reply index.js]: oh, I think it's refering to the way config is passed into secret-stack.
there are some options that are passed in when the SecretStack object is initialized.
https://github.com/ssbc/scuttlebot/blob/master/index.js#L165-L171
and then the rest is passed in when the sbot instance is created: https://github.com/ssbc/scuttlebot/blob/master/bin.js#L61

Really, it would be cleaner to just do the second one... but removing the first thing would break the tests which is why I havn't fixed it yet. (i think)

**Comment by dominictarr** [comment]
is there a description of what the multiserver config structure looks like somewhere?

**Comment by arj03** [comment]
@dominictarr you mean this? https://github.com/ssbc/ssb-config/pull/13

**Comment by cryptix** [comment]
Hmm... the scuttlebot branch/PR I prepared that uses this branch, runs into [this](https://travis-ci.org/ssbc/scuttlebot/jobs/424511802#L1455) crash on node@6 [but not node@8](https://travis-ci.org/ssbc/scuttlebot/builds/424511800).. I have no idea why it does that yet...


```
/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/index.js:219
          ;(server.close || server)(function (err) {
                  ^
TypeError: Cannot read property 'close' of undefined
    at close (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/index.js:219:19)
    at hooked (/home/travis/build/ssbc/scuttlebot/node_modules/hoox/index.js:10:15)
    at /home/travis/build/ssbc/scuttlebot/index.js:60:9
    at next (/home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/index.js:169:9)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
npm ERR! Test failed.  See above for more details.
```

**Comment by cryptix** [comment]
> is there a description of what the multiserver config structure looks like somewhere?

@dominictarr: wasn't 100% sure where to add that. Might make more sense to describe it on the multisrv readme and link to it from ssb-configs readme?

**Comment by dominictarr** [comment]
@cryptix the only thing I can think of is that it's trying to close the server before it's actually been created yet.

@arj03 yes, thanks that was what I wanted to see.

**Comment by dominictarr** [comment]
@cryptix there is something weird about setImmediate on v6. on a hunch, i changed that to `process.nextTick`, and it worked. This only affects the plugin test, which also closes the server immediately. But that got another crash... and then i discovered a less ugly solution that fixed both!
